storm-core/src/jvm/org/apache/storm/daemon/Acker.java

```diff
@@ -0,0 +1,127 @@
+
+import org.apache.storm.task.IBolt;
+import org.apache.storm.task.OutputCollector;
+import org.apache.storm.task.TopologyContext;
+import org.apache.storm.tuple.Tuple;
+import org.apache.storm.tuple.Values;
+import org.apache.storm.utils.RotatingMap;
+import org.apache.storm.utils.TupleUtils;
+import org.apache.storm.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+
+public class Acker implements IBolt {
+    private static final Logger LOG = LoggerFactory.getLogger(Acker.class);
+
+    private static final long serialVersionUID = 4430906880683183091L;
+
+    public static final String ACKER_COMPONENT_ID = "__acker";
+    public static final String ACKER_INIT_STREAM_ID = "__ack_init";
+    public static final String ACKER_ACK_STREAM_ID = "__ack_ack";
+    public static final String ACKER_FAIL_STREAM_ID = "__ack_fail";
+
+    public static final int TIMEOUT_BUCKET_NUM = 3;
+
+    private OutputCollector collector;
+    private RotatingMap<Object, AckObject> pending;
+
+    private class AckObject {
+        public long val = 0L;
+        public Integer spoutTask = null;
+        public boolean failed = false;
+
+        // val xor value
+        public void updateAck(Long value) {
+            val = Utils.bitXor(val, value);
+        }
+    }
+
+    @Override
+    public void prepare(Map stormConf, TopologyContext context, OutputCollector collector) {
+        this.collector = collector;
+        this.pending = new RotatingMap<Object, AckObject>(TIMEOUT_BUCKET_NUM);
+    }
+
+    @Override
+    public void execute(Tuple input) {
+        if (TupleUtils.isTick(input)) {
+            Map<Object, AckObject> tmp = pending.rotate();
+            LOG.debug("Number of timeout tuples:{}", tmp.size());
+            return;
+        }
+
+        String streamId = input.getSourceStreamId();
+        Object id = input.getValue(0);
+        AckObject curr = pending.get(id);
+        if (ACKER_INIT_STREAM_ID.equals(streamId)) {
+            if (curr == null) {
+                curr = new AckObject();
+                curr.val = input.getLong(1);
+                curr.spoutTask = input.getInteger(2);
+                pending.put(id, curr);
+            } else {
+                // If receiving bolt's ack before the init message from spout, just update the xor value.
+                curr.updateAck(input.getLong(1));
+                curr.spoutTask = input.getInteger(2);
+            }
+        } else if (ACKER_ACK_STREAM_ID.equals(streamId)) {
+            if (curr != null) {
+                curr.updateAck(input.getLong(1));
+            } else {
+                curr = new AckObject();
+                curr.val = input.getLong(1);
+                pending.put(id, curr);
+            }
+        } else if (ACKER_FAIL_STREAM_ID.equals(streamId)) {
+            if (curr == null) {
+                // The tuple has been already timeout or failed. So, do nothing
```

Comment:
Just for my understanding, it is better that just return here to ensure only one failure for a certain tuple, rather than sending a failure back to spout. Since it is difficult to tell what the real root cause is here(FAIL occurs before INIT, or tuple has been already timeout or failed), if we just send a failure back here, for the latter scenarios, acker will send more than one message back to spout. It will cause incorrect stats of emitting tuple of acker, and potential duplicated calling of fail() in spout (it depends on the implementation of spout).